flink-state-backends/flink-statebackend-heap-spillable/src/main/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMap.java

```diff
@@ -0,0 +1,1527 @@
+		int offsetInChunk = SpaceUtils.getChunkOffsetByAddress(node);
+		ByteBuffer byteBuffer = chunk.getByteBuffer(offsetInChunk);
+		int offsetInByteBuffer = chunk.getOffsetInByteBuffer(offsetInChunk);
+
+		int level = SkipListUtils.getLevel(byteBuffer, offsetInByteBuffer);
+		int keyDataLen = SkipListUtils.getKeyLen(byteBuffer, offsetInByteBuffer);
+		int keyDataOffset = offsetInByteBuffer + SkipListUtils.getKeyDataOffset(level);
+
+		K key = skipListKeySerializer.deserializeKey(byteBuffer, keyDataOffset, keyDataLen);
+		N namespace = skipListKeySerializer.deserializeNamespace(byteBuffer, keyDataOffset, keyDataLen);
+		long valuePointer = SkipListUtils.getValuePointer(byteBuffer, offsetInByteBuffer);
+		S state = helpGetState(valuePointer);
+
+		return new StateEntry.SimpleStateEntry<>(key, namespace, state);
+	}
+
+	// ----------------------------------------------------------------------------------
+
+	@Override
+	public Stream<K> getKeys(N namespace) {
+		updateStat();
+		byte[] namespaceBytes = skipListKeySerializer.serializeNamespace(namespace);
+		ByteBuffer namespaceByteBuffer = ByteBuffer.wrap(namespaceBytes);
+		Iterator<Long> nodeIter = new NamespaceNodeIterator(namespaceByteBuffer, 0, namespaceBytes.length);
+		return StreamSupport.stream(Spliterators.spliteratorUnknownSize(nodeIter, 0), false)
+			.map(this::helpGetKey);
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public int sizeOfNamespace(Object namespace) {
+		updateStat();
+		byte[] namespaceBytes = skipListKeySerializer.serializeNamespace((N) namespace);
+		ByteBuffer namespaceByteBuffer = ByteBuffer.wrap(namespaceBytes);
+		Iterator<Long> nodeIter = new NamespaceNodeIterator(namespaceByteBuffer, 0, namespaceBytes.length);
+		int size = 0;
+		while (nodeIter.hasNext()) {
+			nodeIter.next();
+			size++;
+		}
+
+		return size;
+	}
+
+	@Nonnull
+	@Override
+	public Iterator<StateEntry<K, N, S>> iterator() {
+		updateStat();
+		final Iterator<Long> nodeIter = new NodeIterator();
+		return new Iterator<StateEntry<K, N, S>>() {
+			@Override
+			public boolean hasNext() {
+				return nodeIter.hasNext();
+			}
+
+			@Override
+			public StateEntry<K, N, S> next() {
+				return helpGetStateEntry(nodeIter.next());
+			}
+		};
+	}
+
+	@Override
+	public InternalKvState.StateIncrementalVisitor<K, N, S> getStateIncrementalVisitor(int recommendedMaxNumberOfReturnedRecords) {
+		return new StateIncrementalVisitor(recommendedMaxNumberOfReturnedRecords);
+	}
+
+	@Nonnull
+	@Override
+	public StateMapSnapshot<K, N, S, ? extends StateMap<K, N, S>> stateSnapshot() {
+		tryToDeleteNodesPhysically();
+
+		ResourceGuard.Lease lease;
+		try {
+			lease = resourceGuard.acquireResource();
+		} catch (Exception e) {
+			throw new RuntimeException("Acquire resource failed, and can't make snapshot of state map", e);
+		}
+
+		synchronized (snapshotVersions) {
+			// increase the map version for copy-on-write and register the snapshot
+			if (++stateMapVersion < 0) {
```

Comment:
This method is a blocking call invoked by main thread (the synchronous part even the checkpoint is configured as asynchronous, we must wait until snapshot (of the meta) complete before accepting new writes, only the physical-data-flush and upload-to-DFS phase could be asynchronous). However, since snapshot is a low frequency operation, I agree there's no big harm making `stateMapVersion` volatile.